### PR TITLE
PW-2525 - Update 3ds2 flow

### PR DESCRIPTION
**Pull request by martinsrenato**
- Create PENDING order before redirect
- Restore order if payment not authorised
- Unit tests

### CODE REVIEW

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java [hunk 1]
```diff
@@ -254,25 +255,9 @@ public String placeOrder(@ModelAttribute("placeOrderForm") final PlaceOrderForm
...
-                    if (adyenPaymentMethod.equals(PAYMENT_METHOD_CC) || adyenPaymentMethod.indexOf(PAYMENT_METHOD_ONECLICK) == 0) {
-                        model.addAttribute(MODEL_CHECKOUT_SHOPPER_HOST, adyenCheckoutFacade.getCheckoutShopperHost());
-                        model.addAttribute(MODEL_ENVIRONMENT_MODE, adyenCheckoutFacade.getEnvironmentMode());
-                        model.addAttribute(SHOPPER_LOCALE, adyenCheckoutFacade.getShopperLocale());
-                        model.addAttribute(PAYMENT_DATA, paymentsResponse.getPaymentData());
-                        model.addAttribute(CHALLENGE_TOKEN, paymentsResponse.getAuthentication().get(THREEDS2_CHALLENGE_TOKEN));
-                        return AdyenControllerConstants.Views.Pages.MultiStepCheckout.Validate3DS2PaymentPage;
-                    }
+                if ((CHALLENGESHOPPER == paymentsResponse.getResultCode() || IDENTIFYSHOPPER == paymentsResponse.getResultCode())
+                        && is3DSPaymentMethod(adyenPaymentMethod)) {
```
[the remark this answers is not in the document]
Comment (martinsrenato) [reply adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java:259]: I also thought that maybe it's not needed, but this code was already there on the previous version so I kept it (I just created a method to make it more readable).

@rkewlani do you think there's any need for this?

[the remark this answers is not in the document]
Comment (martinsrenato) [reply adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java]: Fixed

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java [hunk 2]
```diff
@@ -660,27 +613,40 @@ public OrderData handle3DS2Response(final HttpServletRequest request) throws Exc
...
-                return createAuthorizedOrder(paymentsResponse);
-            }
-            throw new AdyenNonAuthorizedPaymentException(paymentsResponse);
-        } catch (ApiException e) {
+            paymentsResponse = getAdyenPaymentService().authorise3DS2Payment(paymentData, token, type);
+        } catch (Exception e) {
+            LOGGER.debug(e instanceof ApiException ? e.toString() : e.getMessage());
             if (type.equals("challenge")) {
                 LOGGER.debug("Restoring cart because ApiException occurred after challengeResult ");
-                restoreSessionCart();
```
Comment (rkewlani) [inline adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:676]: can we clean up restoreSessionCart() ?

**Comment by rkewlani** [comment]
Since it is at multiple places, hence adding comments here: can we cleanup old locksessioncart() and restoresessioncart() methods since we are not locking cart in session anymore?

**Comment by martinsrenato** [comment]
> Since it is at multiple places, hence adding comments here: can we cleanup old locksessioncart() and restoresessioncart() methods since we are not locking cart in session anymore?

@rkewlani these methods are still used on flow we haven't touched, that's why I haven't completely removed it.
One flow is Paypal component, that we'll soon fix, so wouldn't be bad to temporally break it on the feature branch.
The other is Paypal ECS:do we plan to change it? Or remove it?

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java (hunk 2 above)
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:676]: answered on comment below

**Pull request merged**